File: ios/EuvicMobile.swift
//
//  EuvicMobileModule.swift
//  EuvicMobileSDK
//
//  Created by Kamil Modzelewski on 21/10/2022.
//  Copyright © 2022 SPEEDNET. All rights reserved.
//
//

import Foundation
import EuvicMobileSDK

@objc(EuvicMobile)
internal class EuvicMobile: NSObject {
  public override init() {
    super.init()
  }
  
  @objc public func getCurrentUserId(_ callback: RCTResponseSenderBlock) {
    callback([EuvicMobileSDK.EuvicMobile.shared.currentUserId])
  }
  
  @objc public func configure(_ url: String, apiKey: String, userId: String? = nil, currency: String? = nil, allowSensitiveData: Bool = true) {
    DispatchQueue.main.async {
      EuvicMobileSDK.EuvicMobile.shared.config = .init(
        url: url,
        apiKey: apiKey,
        userId: userId,
        currency: currency,
        allowSensitiveData: allowSensitiveData
      )
    }
  }
  
  @objc public func homepageVisitedEvent(_ custom: [String: Any] = [:]) {
    DispatchQueue.main.async {
      EuvicMobileSDK.EuvicMobile.shared.homepageVisitedEvent(custom: custom)
    }
  }
  
  @objc public func productBrowsedEvent(_ product: [String: Any], custom: [String: Any] = [:]) {
      guard
          let id = product["id"] as? String,
          let price = product["price"] as? String,
          let quantity = product["quantity"] as? Int
      else {
          return
      }
      
      let currency = product["currency"] as? String
      let product = EuvicMobileSDK.EuvicMobileProduct(id: id, price: price, currency: currency, quantity: quantity)
      
      EuvicMobileSDK.EuvicMobile.shared.productBrowsedEvent(
          product: product,
          custom: custom
      )
  }
  
  @objc public func productAddedEvent(_ product: [String: Any], custom: [String: Any] = [:]) {
    guard
        let id = product["id"] as? String,
        let price = product["price"] as? String,
        let quantity = product["quantity"] as? Int
    else {
        return
    }
    
    let currency = product["currency"] as? String
    let product = EuvicMobileSDK.EuvicMobileProduct(id: id, price: price, currency: currency, quantity: quantity)
    
    DispatchQueue.main.async {
      EuvicMobileSDK.EuvicMobile.shared.productAddedEvent(
        product: product,
        custom: custom
      )
    }
  }
  
  @objc public func productRemovedEvent(_ product: [String: Any], custom: [String: Any] = [:]) {
    guard
        let id = product["id"] as? String,
        let price = product["price"] as? String,
        let quantity = product["quantity"] as? Int
    else {
        return
    }
    
    let currency = product["currency"] as? String
    let product = EuvicMobileSDK.EuvicMobileProduct(id: id, price: price, currency: currency, quantity: quantity)
    
    DispatchQueue.main.async {
      EuvicMobileSDK.EuvicMobile.shared.productRemovedEvent(
        product: product,
        custom: custom
      )
    }
  }
  
  @objc public func browsedCategoryEvent(_ name: String, products: [[String: Any]], custom: [String: Any] = [:]) {
    var items = [EuvicMobileSDK.EuvicMobileProduct]()
    for product in products {
      guard
          let id = product["id"] as? String,
          let price = product["price"] as? String,
          let quantity = product["quantity"] as? Int
      else {
          continue
      }
      
      let currency = product["currency"] as? String
      items.append(EuvicMobileSDK.EuvicMobileProduct(id: id, price: price, currency: currency, quantity: quantity))
    }
    
    DispatchQueue.main.async {
      EuvicMobileSDK.EuvicMobile.shared.browsedCategoryEvent(
        name: name,
        products: items,
        custom: custom
      )
    }
  }
  
  @objc public func cartEvent(_ products: [[String: Any]], custom: [String: Any] = [:]) {
    var items = [EuvicMobileSDK.EuvicMobileProduct]()
    for product in products {
      guard
          let id = product["id"] as? String,
          let price = product["price"] as? String,
          let quantity = product["quantity"] as? Int
      else {
          continue
      }
      
      let currency = product["currency"] as? String
      items.append(EuvicMobileSDK.EuvicMobileProduct(id: id, price: price, currency: currency, quantity: quantity))
    }
    
    DispatchQueue.main.async {
      EuvicMobileSDK.EuvicMobile.shared.cartEvent(
        products: items,
        custom: custom
      )
    }
  }
  
  @objc public func orderStartedEvent(_ custom: [String: Any] = [:]) {
    DispatchQueue.main.async {
      EuvicMobileSDK.EuvicMobile.shared.orderStartedEvent(
        custom: custom
      )
    }
  }
  
  @objc public func productsOrderedEvent(_ orderId: String, saleValue: String, products:[[String: Any]], currency: String? = nil, custom: [String: Any] = [:]) {
    var items = [EuvicMobileSDK.EuvicMobileProduct]()
    for product in products {
      guard
          let id = product["id"] as? String,
          let price = product["price"] as? String,
          let quantity = product["quantity"] as? Int
      else {
          continue
      }
      
      let currency = product["currency"] as? String
      items.append(EuvicMobileSDK.EuvicMobileProduct(id: id, price: price, currency: currency, quantity: quantity))
    }
    
    DispatchQueue.main.async {
      EuvicMobileSDK.EuvicMobile.shared.productsOrderedEvent(
        orderId: orderId,
        saleValue: saleValue,
        products: items,
        currency: currency,
        custom: custom
      )
    }
  }
  
  @objc internal static func requiresMainQueueSetup() -> Bool {
    return false
  }
}

File: android/app/src/main/java/com/euvicsdkreactnativedemo/EuvicMobile.java
package com.euvicsdkreactnativedemo;

import android.content.Context;

import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import pl.speednet.euvictrackersdk.EuvicMobileSDK;
import pl.speednet.euvictrackersdk.events.CustomParams;
import pl.speednet.euvictrackersdk.events.Product;

public class EuvicMobile extends ReactContextBaseJavaModule {

    EuvicMobile(ReactApplicationContext context) {
        super(context);
    }

    @Override
    public String getName() {
        return "EuvicMobile";
    }

    EuvicMobileSDK sdk = EuvicMobileSDK.INSTANCE;

    @ReactMethod
    public void configure(
            String url,
            String apiKey,
            String userId,
            String currency,
            Boolean allowSensitiveData
    ) {
        Context context = getReactApplicationContext();

        sdk.configure(
                context,
                url,
                apiKey,
                userId,
                currency,
                allowSensitiveData

        );
    }

    @ReactMethod
    public void getCurrentUserId(Callback callBack) {
        try {
            String userId = sdk.getGetCurrentUserId();
            callBack.invoke(userId);
        } catch (Exception ignored) {

        }
    }

    @ReactMethod
    public void homepageVisitedEvent(ReadableMap custom) {
        sdk.homepageVisitedEvent(mapToCustomParams(custom));
    }

    @ReactMethod
    public void productBrowsedEvent(ReadableMap product, ReadableMap custom) {
        Product productObject = mapToProduct(product.toHashMap());
        if (productObject != null) {
            sdk.productBrowsedEvent(
                    productObject,
                    mapToCustomParams(custom)
            );
        }
    }

    @ReactMethod
    public void productAddedEvent(ReadableMap product, ReadableMap custom) {
        Product productObject = mapToProduct(product.toHashMap());
        if (productObject != null) {
            sdk.productAddedEvent(
                    mapToProduct(product.toHashMap()),
                    mapToCustomParams(custom)
            );
        }
    }

    @ReactMethod
    public void productRemovedEvent(ReadableMap product, ReadableMap custom) {
        Product productObject = mapToProduct(product.toHashMap());
        if (productObject != null) {
            sdk.productRemovedEvent(
                    productObject,
                    mapToCustomParams(custom)
            );
        }
    }

    @ReactMethod
    public void browsedCategoryEvent(String categoryName, ReadableArray products, ReadableMap custom) {
        List<Product> productList = mapToProductList(products.toArrayList());

        if (!productList.isEmpty()) {
            sdk.browsedCategoryEvent(
                    categoryName,
                    productList,
                    mapToCustomParams(custom)
            );
        }
    }

    @ReactMethod
    public void cartEvent(ReadableArray products, ReadableMap custom) {
        List<Product> productList = mapToProductList(products.toArrayList());

        if (!productList.isEmpty()) {
            sdk.cartEvent(
                    productList,
                    mapToCustomParams(custom)
            );
        }
    }

    @ReactMethod
    public void orderStartedEvent(ReadableMap custom) {
        sdk.orderStartedEvent(
                mapToCustomParams(custom)
        );
    }

    @ReactMethod
    public void productsOrderedEvent(String orderId, String saleValue, ReadableArray products, String currency, ReadableMap custom) {
        List<Product> productList = mapToProductList(products.toArrayList());

        sdk.productsOrderedEvent(
                orderId,
                saleValue,
                productList,
                currency,
                mapToCustomParams(custom)
        );
    }

    private List<Product> mapToProductList(List<Object> productsArrayList) {

        List<Product> productList;

        if (productsArrayList != null) {
            ArrayList<Product> mutableList = new ArrayList<>();

            for (Object element : (Iterable<Object>) productsArrayList) {
                HashMap<String, Object> hashProduct = (HashMap) element;

                Product getProduct = mapToProduct((HashMap) hashProduct);

                if (getProduct != null) {
                    mutableList.add(getProduct);
                }
            }
            productList = mutableList;
        } else {
            productList = null;
        }

        return productList;
    }

    private Product mapToProduct(HashMap<String, Object> map) {
        try {

            Object id = map.get("id");
            if (!(id instanceof String)) {
                id = null;
            }

            Object price = map.get("price");
            if (!(price instanceof String)) {
                price = null;
            }

            Object currency = map.get("currency");
            if (!(currency instanceof String)) {
                currency = null;
            }

            Object quantity = map.get("quantity");
            if (!(quantity instanceof Integer)) {
                if (quantity instanceof Double) {
                    long longQuantity = Math.round((Double) quantity);
                    quantity = Math.toIntExact(longQuantity);
                } else if (quantity instanceof Long) {
                    quantity = Math.toIntExact((Long) quantity);
                } else {
                    quantity = null;
                }
            }


            Product product;
            if (id != null && price != null && quantity != null) {

                if (currency != null) {
                    product = new Product(id.toString(), price.toString(), currency.toString(), (int) quantity);
                } else {
                    product = new Product(id.toString(), price.toString(), null, (int) quantity);
                }

            } else {
                product = null;
            }
            return product;
        } catch (Exception e) {

        }
        return null;
    }

    private Function1 mapToCustomParams(final ReadableMap map) {
        return map == null || map.toHashMap().isEmpty() ? null : (Function1) (new Function1() {
            public Object invoke(Object object) {
                this.invoke((CustomParams) object);
                return Unit.INSTANCE;
            }

            public void invoke(@NotNull CustomParams receiver) {
                Map<String, Object> hashMap = map.toHashMap();

                for (Map.Entry<String, Object> stringObjectEntry : hashMap.entrySet()) {

                    String key = (String) ((Map.Entry<String, Object>) stringObjectEntry).getKey();
                    Object value = ((Map.Entry<String, Object>) stringObjectEntry).getValue();

                    if (value instanceof String) {
                        receiver.param(key, (String) value);
                    } else if (value instanceof Integer) {
                        receiver.param(key, ((Number) value).intValue());
                    } else if (value instanceof Float) {
                        receiver.param(key, ((Number) value).floatValue());
                    } else if (value instanceof Double) {
                        receiver.param(key, ((Number) value).doubleValue());
                    } else if (value instanceof Boolean) {
                        receiver.param(key, ((Boolean) value).toString());
                    }
                }

            }
        });
    }

}